gestalt/vault.py

```diff
@@ -243,3 +246,41 @@ def worker(self, token_queue: Queue) -> None:  # type: ignore
     @property
     def scheme(self) -> str:
         return self._scheme
+
+    def _validate_token_expiration(self) -> None:
+        token_details = self.vault_client.auth.token.lookup_self()
+        if token_details['data'] is not None:
+
+            expire_time = None
+            if 'expire_time' not in token_details['data']:
+                print(
+                    "Key 'expire_time' does not exist in token_details['data']"
+                )
+                return None
+
+            expire_time = token_details['data']['expire_time']
+            # Validate expire_time is present
+            if expire_time is None:
+                print("Cannot parse expire_time, value is None")
+                return None
+
+            # Use isoparse to correctly parse the datetime string
+            expire_time = isoparse(expire_time)
+
+            # Ensure the parsed time is in UTC
+            if expire_time.tzinfo is None:
+                expire_time = expire_time.replace(tzinfo=timezone.utc)
+            else:
+                expire_time = expire_time.astimezone(timezone.utc)
+
+            threshold = timedelta(days=EXPIRATION_THRESHOLD_DAYS)
+            current_time = datetime.now(timezone.utc)
+            delta_time = expire_time - current_time
+
+            if delta_time <= threshold:
+                print(f"Re-auth with vault.")
+                self.connect()
```

Comment:
this is going to cause additional threads to be created, right? I think we need to stop our existing threads before creating new ones. This sounds like a memory leak.

Reply:
Being handled.